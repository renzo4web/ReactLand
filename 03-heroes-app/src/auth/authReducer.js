import { types } from '../types/types';

const authReducer = (state = {}, action) => {
  const { type, payload } = action;

  switch (type) {
    case types.login:
      return { ...state, ...payload, logged: true };

    case types.logout:
      return { ...payload, logged: false };
    case 'default':
      return state;

    default:
      throw new Error(
        `authReducer receibed invalid type ${type}`
      );
  }
};

export { authReducer };
