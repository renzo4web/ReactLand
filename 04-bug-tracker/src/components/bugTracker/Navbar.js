import React from "react";
import { NavLink } from "react-router-dom";

const Navbar = () => {
  return (
    <div className="navbar__container">
      <nav className="navbar">
        <ul className="navbar__list">
          <li>
            <NavLink className="navbar__item" to="/">
              Dashboard
            </NavLink>
          </li>
          <li>
            <NavLink clasname="navbar__item" to="/bugs">
              Bugs
            </NavLink>
          </li>
        </ul>

        <div className="navbar__list">
          <button className="ui__btn-logout">Logout</button>
        </div>
      </nav>
    </div>
  );
};

export default Navbar;
