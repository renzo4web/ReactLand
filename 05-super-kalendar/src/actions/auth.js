import { types } from '../types/types';
import { fetchWithoutToken, fetchWithToken } from '../helpers/fetch';
import toast from 'react-hot-toast';
import { eventsCleanAll } from './events';

export const startLogin = (email, password) => {
	return async dispatch => {
		// http://localhost:4000/api/auth
		const response = await fetchWithoutToken(
			'auth',
			{ email, password },
			'POST'
		);
		const body = await response.json();

		if (body.ok) {
			localStorage.setItem('token', body.token);
			// keep track initial date of the token created
			localStorage.setItem('token-init-date', new Date().getTime());

			const { uid, name } = body;

			dispatch(
				login({
					uid,
					name,
				})
			);
		} else {
			toast.error(body.msg);
		}
	};
};

export const startRegister = (name, email, password) => {
	return async dispatch => {
		const response = await fetchWithoutToken(
			'auth/new',
			{ email, password, name },
			'POST'
		);

		const body = await response.json();

		if (body.ok) {
			localStorage.setItem('token', body.token);
			// keep track initial date of the token created
			localStorage.setItem('token-init-date', new Date().getTime());

			const { uid, name } = body;

			dispatch(
				login({
					uid,
					name,
				})
			);
		} else {
			toast.error(body.msg);
		}
	};
};

export const startChecking = () => {
	return async dispatch => {
		const response = await fetchWithToken('auth/renew');
		const body = await response.json();

		if (body.ok) {
			localStorage.setItem('token', body.token);
			// keep track initial date of the token created
			localStorage.setItem('token-init-date', new Date().getTime());

			const { uid, name } = body;

			dispatch(
				login({
					uid,
					name,
				})
			);
		} else {
			dispatch(checkingFinish());
		}
	};
};

const checkingFinish = () => ({
	type: types.authCheckingFinish,
});

export const startLogout = () => {
	return dispatch => {
		localStorage.clear();
		dispatch(logout());
		dispatch(eventsCleanAll());
	};
};

const logout = () => ({
	type: types.authLogout,
});

const login = user => ({
	type: types.authLogin,
	payload: user,
});
