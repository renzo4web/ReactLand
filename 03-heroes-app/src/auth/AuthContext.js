import React, {
  createContext,
  useReducer,
  useContext,
  useEffect,
} from 'react';
import { authReducer } from './authReducer';

const Context = createContext();

const init = () => {
  return (
    JSON.parse(localStorage.getItem('user')) || {
      logged: false,
    }
  );
};

const AuthContext = ({ children }) => {
  const value = useReducer(authReducer, {}, init);
  const [auth, dispatch] = value;

  useEffect(() => {
    localStorage.setItem('user', JSON.stringify(auth));
  }, [auth]);

  return (
    <Context.Provider value={value}>
      {children}
    </Context.Provider>
  );
};

export const useAuthContext = () => {
  return useContext(Context);
};

export { AuthContext };
