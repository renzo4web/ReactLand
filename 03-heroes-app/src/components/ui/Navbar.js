import React from 'react';
import {
  Box,
  Flex,
  Text,
  Button,
  Stack,
  useColorModeValue,
} from '@chakra-ui/react';
import {
  Link,
  useHistory,
  useLocation,
} from 'react-router-dom';
import { ColorModeSwitcher } from '../../ColorModeSwitcher';
import Search from '../search/Search';
import { useAuthContext } from '../../auth/AuthContext';
import { types } from '../../types/types';

export const Navbar = props => {
  const { name } = useAuthContext()[0];
  const [isOpen, setIsOpen] = React.useState(false);

  const toggle = () => setIsOpen(!isOpen);

  return (
    <NavBarContainer {...props}>
      <MenuToggle toggle={toggle} isOpen={isOpen} />
      <MenuLinks isOpen={isOpen} />
      <Text
        fontSize="xl"
        as="strong"
        ml="auto"
        fontWeight="bold"
        color="white"
      >
        {name}
      </Text>
      <ColorModeSwitcher />
    </NavBarContainer>
  );
};

const CloseIcon = () => (
  <svg
    width="24"
    viewBox="0 0 18 18"
    xmlns="http://www.w3.org/2000/svg"
  >
    <title>Close</title>
    <path
      fill="white"
      d="M9.00023 7.58599L13.9502 2.63599L15.3642 4.04999L10.4142 8.99999L15.3642 13.95L13.9502 15.364L9.00023 10.414L4.05023 15.364L2.63623 13.95L7.58623 8.99999L2.63623 4.04999L4.05023 2.63599L9.00023 7.58599Z"
    />
  </svg>
);

const MenuIcon = () => (
  <svg
    width="24px"
    viewBox="0 0 20 20"
    xmlns="http://www.w3.org/2000/svg"
    fill="white"
  >
    <title>Menu</title>
    <path d="M0 3h20v2H0V3zm0 6h20v2H0V9zm0 6h20v2H0v-2z" />
  </svg>
);

const MenuToggle = ({ toggle, isOpen }) => {
  return (
    <Box
      display={{ base: 'block', md: 'none' }}
      onClick={toggle}
    >
      {isOpen ? <CloseIcon /> : <MenuIcon />}
    </Box>
  );
};

const MenuItem = ({
  children,
  isLast,
  to = '/',
  color,
  ...rest
}) => {
  return (
    <Link to={to}>
      <Text
        fontWeight="bold"
        display="block"
        color={color}
        {...rest}
      >
        {children}
      </Text>
    </Link>
  );
};

const MenuLinks = ({ isOpen, color }) => {
  const [auth, dispatch] = useAuthContext();
  const history = useHistory();

  const handleLogout = () => {
    dispatch({
      type: types.logout,
    });
    history.replace('/login');
  };

  return (
    <Box
      display={{
        base: isOpen ? 'block' : 'none',
        md: 'block',
      }}
      flexBasis={{ base: '100%', md: 'auto' }}
    >
      <Stack
        spacing={8}
        color={color}
        align="center"
        justify={[
          'center',
          'space-between',
          'flex-end',
          'flex-end',
        ]}
        direction={['column', 'row', 'row', 'row']}
        pt={[4, 4, 0, 0]}
      >
        <MenuItem to="/">All</MenuItem>
        <MenuItem to="/dc">DC</MenuItem>
        <MenuItem to="/marvel">Marvel</MenuItem>
        <Button
          onClick={handleLogout}
          size="md"
          rounded="md"
          color={color}
          bg={[
            'white',
            'white',
            'primary.500',
            'primary.500',
          ]}
          _hover={{
            bg: [
              'primary.100',
              'primary.100',
              'primary.600',
              'primary.600',
            ],
          }}
        >
          Logout
        </Button>
        <Search />
      </Stack>
    </Box>
  );
};

const NavBarContainer = ({ children, ...props }) => {
  const { pathname } = useLocation();
  const bg = pathname.includes('dc')
    ? 'blue.500'
    : 'red.500';

  const color = useColorModeValue('black', 'gray.800');

  return (
    <Flex
      as="nav"
      align="center"
      justify="space-between"
      wrap="wrap"
      w="100%"
      mb={8}
      p={8}
      bg={bg}
      color={color}
      {...props}
    >
      {children}
    </Flex>
  );
};
