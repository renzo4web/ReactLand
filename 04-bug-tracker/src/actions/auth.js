import Swal from 'sweetalert2';

import { firebase, googleAuthProvider } from '../firebase/firebase-config';
import { types } from '../types/types';
import { bugsLogout } from './bugs';
import { finishLoading, startLoading } from './ui';

export const startLoginEmailPassword = (email, password) => {
    return (dispatch) => {
        dispatch(startLoading());

        firebase
            .auth()
            .signInWithEmailAndPassword(email, password)
            .then(({ user }) => {
                dispatch(login(user.uid, user.displayName));
                dispatch(finishLoading());
            })
            .catch((err) => {
                dispatch(finishLoading());
                Swal.fire('Error', err.message);
            });
    };
};

export const startGoogleLogin = () => {
    return (dispatch) => {
        firebase
            .auth()
            .signInWithPopup(googleAuthProvider)
            .then(({ user }) => {
                const { displayName, uid } = user;
                dispatch(login(uid, displayName));
            });
    };
};

export const startRegisterWithEmailPasswordName = (name, email, password) => {
    return (dispatch) => {
        firebase
            .auth()
            .createUserWithEmailAndPassword(email, password)
            .then(async ({ user }) => {
                await user.updateProfile({
                    displayName: name,
                });

                dispatch(login(user.id, user.displayName));
            })
            .catch((err) => {
                Swal.fire('Error', err.message);
            });
    };
};

export const login = (uid, displayName) => ({
    type: types.login,
    payload: {
        uid,
        displayName,
    },
});

export const startLogout = () => {
    return async (dispatch) => {
        await firebase.auth().signOut();
        dispatch(bugsLogout());
        dispatch(logout());
    };
};

export const logout = () => ({
    type: types.logout,
});
