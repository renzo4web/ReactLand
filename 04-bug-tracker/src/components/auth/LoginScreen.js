import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import validator from 'validator';
import { Link } from 'react-router-dom';
import { startGoogleLogin, startLoginEmailPassword } from '../../actions/auth';
import { useForm } from '../../hooks/useForm';
import { removeError, setError } from '../../actions/ui';

const LoginScreen = () => {
    const initialForm = {
        email: '',
        password: '',
    };
    const { msgError, loading } = useSelector((state) => state.ui);
    const dispatch = useDispatch();
    const [formValues, handleInputChange] = useForm(initialForm);

    const { email, password } = formValues;

    const handleLogin = (e) => {
        e.preventDefault();
        if (isFormValid()) {
            dispatch(startLoginEmailPassword(email, password));
        }
    };

    const handleGoogleLogin = () => {
        dispatch(startGoogleLogin());
    };

    const isFormValid = () => {
        if (!validator.isEmail(email)) {
            dispatch(setError('Email is not valid'));
            return false;
        }

        if (password.length < 5) {
            dispatch(
                setError(
                    'Password should be at least 5 characters long and match each other'
                )
            );
            return false;
        }

        dispatch(removeError());
        return true;
    };

    return (
        <>
            <h1>Login</h1>

            {msgError && (
                <strong className='ui__alert-input'>{msgError}</strong>
            )}
            <form onSubmit={handleLogin}>
                <div className='ui__input-group'>
                    <label htmlFor='email'>Email</label>
                    <input
                        value={email}
                        placeholder='test@gmail.com'
                        onChange={handleInputChange}
                        name='email'
                        id='email'
                        type='text'
                    />
                </div>

                <div className='ui__input-group'>
                    <label htmlFor='password'>Password</label>
                    <input
                        value={password}
                        onChange={handleInputChange}
                        name='password'
                        id='password'
                        type='password'
                    />
                </div>

                <button disabled={loading} type='submit' className='auth__btn'>
                    Login
                </button>

                <hr />

                <div>
                    <p>Login with social networks</p>

                    <div onClick={handleGoogleLogin} className='google-btn'>
                        <div className='google-icon-wrapper'>
                            <img
                                className='google-icon'
                                src='https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg'
                                alt='google button'
                            />
                        </div>
                        <p className='btn-text'>
                            <b>Sign in with google</b>
                        </p>
                    </div>
                </div>

                <Link to='/auth/register'>Create new account</Link>
            </form>
        </>
    );
};

export default LoginScreen;
