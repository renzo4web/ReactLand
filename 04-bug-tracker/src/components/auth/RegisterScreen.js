import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { Link } from 'react-router-dom';
import validator from 'validator';

import {
    registerFirestore,
    startRegisterWithEmailPasswordName,
} from '../../actions/auth';
import { removeError, setError } from '../../actions/ui';
import { useForm } from '../../hooks/useForm';

const RegisterScreen = () => {
    const { msgError } = useSelector((state) => state.ui);
    const dispatch = useDispatch();
    console.log(msgError);

    const [formValues, handleInputChange] = useForm({
        name: 'TEst',
        email: 'test@gmail.com',
        password: '123456',
        password2: '123456',
    });

    const { name, email, password, password2 } = formValues;

    const handleSubmit = (e) => {
        e.preventDefault();

        if (isFormValid()) {
            dispatch(startRegisterWithEmailPasswordName(name, email, password));
        }
    };

    const isFormValid = () => {
        if (name.trim().length === 0) {
            dispatch(setError('Name invalid'));
            return false;
        }

        if (!validator.isEmail(email)) {
            dispatch(setError('Email is not valid'));
            return false;
        }

        if (password !== password2 || password.length < 5) {
            dispatch(
                setError(
                    'Password should be at least 5 characters long and match each other'
                )
            );
            return false;
        }

        dispatch(removeError());
        return true;
    };

    return (
        <>
            <h1>Register</h1>

            {msgError && (
                <strong className='ui__alert-input'>{msgError}</strong>
            )}
            <form onSubmit={handleSubmit}>
                <div className='ui__input-group'>
                    <label htmlFor='name'>Name</label>
                    <input
                        onChange={handleInputChange}
                        name='name'
                        id='name'
                        type='text'
                        value={name}
                    />
                </div>

                <div className='ui__input-group'>
                    <label htmlFor='email'>Email</label>
                    <input
                        onChange={handleInputChange}
                        value={email}
                        name='email'
                        id='email'
                        type='text'
                    />
                </div>

                {/* TODO : CHECK THE PASSAWORD ARE THE SAME */}

                <div className='ui__input-group'>
                    <label htmlFor='password'>Password</label>
                    <input
                        onChange={handleInputChange}
                        value={password}
                        name='password'
                        id='password'
                        type='password'
                    />
                </div>

                <div className='ui__input-group'>
                    <label htmlFor='password2'>Confirm password</label>
                    <input
                        onChange={handleInputChange}
                        value={password2}
                        name='password2'
                        id='password2'
                        type='password'
                    />
                </div>
                <button className='ui__button'>Enter</button>

                <hr />

                <div>
                    <p>Register with social networks</p>

                    <div className='google-btn'>
                        <div className='google-icon-wrapper'>
                            <img
                                className='google-icon'
                                src='https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg'
                                alt='google button'
                            />
                        </div>
                        <p className='btn-text'>
                            <b>Sign in with google</b>
                        </p>
                    </div>
                </div>

                <Link className='link' to='/auth/login'>
                    Already register ?
                </Link>
            </form>
        </>
    );
};

export default RegisterScreen;
