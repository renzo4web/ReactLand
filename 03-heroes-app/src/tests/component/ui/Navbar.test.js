import { mount } from 'enzyme';
import { MemoryRouter } from 'react-router-dom';
import { AuthContext } from '../../../auth/AuthContext';
import { Navbar } from '../../../components/ui/Navbar';

describe('<Navbar /> Tests', () => {
    test('should render properly', () => {
        const wrapper = mount(
            <AuthContext>
                <MemoryRouter>
                    <Navbar />
                </MemoryRouter>
            </AuthContext>
        );

        expect(wrapper).toMatchSnapshot();
    });
});
