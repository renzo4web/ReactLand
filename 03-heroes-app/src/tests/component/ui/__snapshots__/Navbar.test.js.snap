// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Navbar /> Tests should render properly 1`] = `
<nav
  className="css-qluv0w"
>
  <div
    className="css-dsyim5"
    onClick={[Function]}
  >
    <svg
      fill="white"
      viewBox="0 0 20 20"
      width="24px"
      xmlns="http://www.w3.org/2000/svg"
    >
      <title>
        Menu
      </title>
      <path
        d="M0 3h20v2H0V3zm0 6h20v2H0V9zm0 6h20v2H0v-2z"
      />
    </svg>
  </div>
  <div
    className="css-8olkqp"
  >
    <div
      className="chakra-stack css-1agnwj4"
    >
      <a
        href="/"
        onClick={[Function]}
      >
        <p
          className="chakra-text css-6vcm9r"
        >
          All
        </p>
      </a>
      <a
        href="/dc"
        onClick={[Function]}
      >
        <p
          className="chakra-text css-6vcm9r"
        >
          DC
        </p>
      </a>
      <a
        href="/marvel"
        onClick={[Function]}
      >
        <p
          className="chakra-text css-6vcm9r"
        >
          Marvel
        </p>
      </a>
      <button
        className="chakra-button css-1xwdflt"
        onClick={[Function]}
        type="button"
      >
        Logout
      </button>
      <form
        onSubmit={[Function]}
      >
        <div
          className="chakra-input__group css-1fzfw9r"
        >
          <div
            className="chakra-input__left-element css-1u2k0bl"
          >
            <svg
              className="chakra-icon css-1xzgr6b"
              focusable={false}
              viewBox="0 0 24 24"
            >
              <path
                d="M23.384,21.619,16.855,15.09a9.284,9.284,0,1,0-1.768,1.768l6.529,6.529a1.266,1.266,0,0,0,1.768,0A1.251,1.251,0,0,0,23.384,21.619ZM2.75,9.5a6.75,6.75,0,1,1,6.75,6.75A6.758,6.758,0,0,1,2.75,9.5Z"
                fill="currentColor"
              />
            </svg>
          </div>
          <input
            className="chakra-input css-1qcmu60"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            placeholder="Search your favorite superhero"
            type="text"
            value=""
          />
        </div>
      </form>
    </div>
  </div>
  <strong
    className="chakra-text css-62oyan"
  />
  <button
    aria-label="Switch to dark mode"
    className="chakra-button css-acoyy6"
    type="button"
  >
    <svg
      aria-hidden={true}
      fill="currentColor"
      focusable={false}
      height="1em"
      stroke="currentColor"
      strokeWidth="0"
      style={
        Object {
          "color": undefined,
        }
      }
      viewBox="0 0 512 512"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M283.211 512c78.962 0 151.079-35.925 198.857-94.792 7.068-8.708-.639-21.43-11.562-19.35-124.203 23.654-238.262-71.576-238.262-196.954 0-72.222 38.662-138.635 101.498-174.394 9.686-5.512 7.25-20.197-3.756-22.23A258.156 258.156 0 0 0 283.211 0c-141.309 0-256 114.511-256 256 0 141.309 114.511 256 256 256z"
        key="0"
      />
    </svg>
  </button>
</nav>
`;
